File: python/pro3.py
# -*- coding: utf-8 -*-

def smallest_factor(n):
    test = 2
    while test < n:
        if 0 == n % test:
            return test
        test = test + 1
    return n

def largest_prime_factor(n):
    num = n
    acc = 1
    while num != 1:
        if num <= 2:
            return num
        else:
            fac = smallest_factor(num)
            num = num / fac
            acc = fac
    return acc

File: ocaml/pro3.ml
let smallest_factor n =
  let rec aux test =
    if test >= n || 0= n mod test then test
    else aux (1+test)
  in aux 2

let largest_prime_factor n =
  let rec aux num acc =
    if 1=num then acc
    else if num <= 2 then num
    else let fac = smallest_factor num
    in aux (num / fac) fac
  in aux n 1
